src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java

```diff
@@ -0,0 +1,609 @@
+      sourceUseLocalSessionState.setValue(true);
+
+      target.setAutoCommit(source.getAutoCommit());
+      final String db = source.getDatabase();
+      if (db != null && !db.isEmpty()) {
+        target.setDatabase(db);
+      }
+      target.setTransactionIsolation(source.getTransactionIsolation());
+      target.setSessionMaxRows(source.getSessionMaxRows());
+
+      sourceUseLocalSessionState.setValue(prevUseLocalSessionState);
+    } catch (SQLException e) {
+      this.logger.logError(Messages.getString("ReadWriteSplittingPlugin.5"));
+      throw new SQLException(
+          Messages.getString("ReadWriteSplittingPlugin.5"),
+          MysqlErrorNumbers.SQL_STATE_UNABLE_TO_CONNECT_TO_DATASOURCE, e);
+    }
+  }
+
+  private synchronized void switchToReaderConnection(JdbcConnection currentConnection) throws SQLException {
+    if (isCurrentConnectionReader() && isConnectionUsable(currentConnection)) {
+      return;
+    }
+
+    if (!isConnectionUsable(this.readerConnection)) {
+      initializeReaderConnection(currentConnection);
+      // The current connection may have changed; update it here in case it did.
+      currentConnection = this.currentConnectionProvider.getCurrentConnection();
+    }
+
+    final int readerHostIndex = getHostIndex(this.readerConnection, this.readerConnection.getHostPortPair());
+    switchCurrentConnectionTo(currentConnection, this.readerConnection, readerHostIndex);
+    this.logger.logDebug(Messages.getString("ReadWriteSplittingPlugin.20",
+            new Object[]{ this.hosts.get(readerHostIndex).getHostPortPair() }));
+  }
+
+  private void initializeReaderConnection(JdbcConnection currentConnection) throws SQLException {
+    if (this.hosts.size() == 1) {
+      if (!isConnectionUsable(this.writerConnection)) {
+        getNewWriterConnection();
+        switchCurrentConnectionTo(currentConnection, this.writerConnection, WRITER_INDEX);
+      }
+      setReaderConnection(this.writerConnection, this.hosts.get(WRITER_INDEX));
+    } else if (this.hosts.size() == 2) {
+      getNewReaderConnection(this.hosts.get(1));
+    } else {
+      final int randomReaderIndex = getRandomReaderIndex();
+      getNewReaderConnection(this.hosts.get(randomReaderIndex));
+    }
+  }
+
+  private int getRandomReaderIndex() {
+    if (this.hosts.size() <= 1) {
+      return NO_CONNECTION_INDEX;
+    }
+
+    final int minReaderIndex = 1;
+    final int maxReaderIndex = this.hosts.size() - 1;
+    return (int) (Math.random() * ((maxReaderIndex - minReaderIndex) + 1)) + minReaderIndex;
+  }
+
+  private boolean isConnectionUsable(JdbcConnection connection) throws SQLException {
+    return connection != null && !connection.isClosed();
+  }
+
+  JdbcConnection getWriterConnection() {
+    return this.writerConnection;
+  }
+
+  JdbcConnection getReaderConnection() {
+    return this.readerConnection;
+  }
+
+  private void logTopology() {
+    StringBuilder msg = new StringBuilder();
```

Comment:
would think we would want to put a guard around this. If logger level less than trace no sense in creating the string

Reply:
Good point, thanks Dave

Reply:
This should have been fixed before closing this PR

Reply:
This PR was closed as Read / Write Splitting has been released with the `aws-advanced-jdbc-wrapper`. As for MySQL support, the `aws-advanced-jdbc-wrapper` is compatible with MySQL and we are currently working towards an official MySQL release in the future.

Reply:
Yes, I figured that out too late.